pkg/vault/contracts/WrappedBalancerPoolToken.sol

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { ERC20Permit } from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Permit.sol";
+
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import { IWrappedBalancerPoolToken } from "@balancer-labs/v3-interfaces/contracts/vault/IWrappedBalancerPoolToken.sol";
+
+/**
+ * @notice ERC20 wrapper for Balancer Pool Token (BPT), allowing users to deposit BPT
+ * and receive 1:1 wrapped tokens, or burn wrapped tokens to redeem the original BPT.
+ * Minting and burning are only allowed when the Vault is locked.
+ */
+contract WrappedBalancerPoolToken is IWrappedBalancerPoolToken, ERC20, ERC20Permit {
+    using SafeERC20 for *;
+
+    IERC20 public immutable bpt;
+    IVault public immutable vault;
+
+    constructor(
+        IVault vault_,
+        IERC20 bpt_,
+        string memory name,
+        string memory symbol
+    ) ERC20(name, symbol) ERC20Permit(name) {
+        vault = vault_;
+        bpt = bpt_;
```

Comment:
Should we check that this is actually a BPT in the Vault? Redundant if it's deployed the intended way (from the factory), but somebody could just deploy it standalone.

Presumably lending protocols would verify that the Wrapped BPT token was deployed by the canonical factory anyway, like exchanges should do now for pools.

```
if (BalancerPoolToken(address(bpt_)).getVault() != vault_) {
    revert TokenIsNotBPT();
}
```

Reply:
you can check whether it's registered here actually

Reply:
(more than `getVault`, just check if initialized here, as `bpt` can also lie about the vault, but not vice versa).

Reply:
I think that if a contract wasn’t deployed through the factory, we can’t trust it at all. Yes, we could verify that the bytecode is a 1:1 match, but in that case, there’s no point in not using the factory for deployment. I’d prefer to keep the factory as the single source of truth for determining which wBPT is canonical.